### ISSUE TITLE
questions about mandarin data packet

### ISSUE THREAD

**Issue by mynah15** [espeak-ng/espeak-ng#1044]
I install espeak for several times, I use the --compile=cmn or zh in dictsource or dictsource/extra , it said 
Full dictionary is not installed for 'zh'
Error processing file 'zh_rules': No such file or directory. 
but there is a file called cmn_rules and cmn_list in dictsource
How can I install mandarin?

And the other question, I try to input chinese and get mandarin phoneme like [input: 你好] output[ n i 3 h ao 3]. Is the espeak  software can output mandarin phoneme or only IPA?
Thanks.

**Comment by jaacoppi** [comment]
Which version are you using? Chinese has changed in current develpoment.

Here's how to see version info:
espeak-ng --version
eSpeak NG text-to-speech: 1.51-dev  Data at: /home/jaacoppi/koodia/espeak-ng/espeak-ng-data
 espeak-ng --version

Short answer:
Looks like mandarin is broken in current development version. Cantonese seems to work as expected. I'll start looking into this.


Long answer:
Chinese used to be divided to zh (Mandarin) and zhy (cantonese). Since commit d91b7cf28 (May 2021) they are called
cmn (Mandarin) and yue (Cantonese).

To support the full dictionaries, configure & compile with:
Mandarin: ./configure --with-extdict-cmn && make
Cantonese: ./configure --with-extdict-cmn && make


There's also a distincion between latin text as pinyin/jyutping and latin text as English.
So, these two should produce different results:
espeak-ng -v cmn "english text 你好 more english text" -x
espeak-ng -v cmn-Latn-pinyin "english text 你好 more english text" -x

**Comment by JuhoHiltunen** [comment]
@valdisvi  @rhdunn  what do you think is the root cause for this?

commit https://github.com/espeak-ng/espeak-ng/commit/48259058358d6d04b3155b1265b2c875489bd301 changes this behavior. However, since Cantonese works I don't want to revert it.

This does not affect all chinese characters. It has something to do with $textmode since ㄦ that's in dictsource/cmn_list before $textmode is set works fine.


There are a few possible solutions:
1) remove "dictrules 1" from espeak-ng-data/lang/sit/cmn (This makes cmn  work like cmn-Latn-pinyin)
2) remove$textmode from dictsource/cmn/list and add it in tr_languages.c (Like Cantonese))


Cantonese simply finds an item in the list, but Mandarin starts to translate it in English:

```
espeak-ng "你" -v yue -X
Found: '你' [nei5]   $text
n'ei5_|

espeak-ng "你" -v cmn -X
Replace: 你   ni3 
Translate 'ni3'
  2     ?1 n     [_^_]
  1     n        [n]
Translate 'ni3'
  1     n        [n]

  1     i        [I]

Found: '_3' [Tr'i:] 




```

**Comment by rhdunn** [comment]
The way that `$textmode` works is that the corresponding text gets placed into the source buffer, so it is as if the text contains those characters. That means that with that change, those entries will now be interpreted as English instead of pinyin.

It would be useful to have a `textmode_dictrules` optin or similar, allowing this to use a different ruleset for that -- i.e. to always use the pinyin rules in this case. That could be quite complex to support.

**Comment by mynah15** [comment]
> 

Thanks for your help.
espeak version is 1.50
espeak-ng -v yue or cmn "你好" worked very will, but 【speak-ng --compile=yue or cmn】in dictsource:
Error processing file 'zhy_rules': No such file or directory.
I change yue to other language like tr tt vi , it can be installed, so i guess is the yue and cmn data packet broken because of change name?

espeak-ng -v cmn "你好" -x
can be read that's make me confuse, as i install language package failed the chinese also can be read.
but the output is n'i35_| x'Au214_|  is that IPA or something else? How can I get Pinyin instead of this?


./configure --with-extdict-cmn && make
Partial output results:
make  all-am
make[1]: Entering directory '/home/bage/Downloads/espeak-ng-master'
  CXXLD    src/espeak-ng
/usr/bin/ld: src/.libs/libespeak-ng.so: undefined reference to `sync_espeak_terminated_msg'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:1227: src/espeak-ng] Error 1
make[1]: Leaving directory '/home/bage/Downloads/espeak-ng-master'
make: *** [Makefile:830: all] Error 2

**Comment by mynah15** [comment]
> Which version are you using? Chinese has changed in current develpoment. Here's how to see version info: espeak-ng --version eSpeak NG text-to-speech: 1.51-dev Data at: /home/jaacoppi/koodia/espeak-ng/espeak-ng-data espeak-ng --version Short answer: Looks like mandarin is broken in current development version. Cantonese seems to work as expected. I'll start looking into this. Long answer: Chinese used to be divided to zh (Mandarin) and zhy (cantonese). Since commit [d91b7cf](https://github.com/espeak-ng/espeak-ng/commit/d91b7cf2889571ff66b7b606129eda79a94a4382) (May 2021) they are called cmn (Mandarin) and yue (Cantonese). To support the full dictionaries, configure & compile with: Mandarin: ./configure --with-extdict-cmn && make Cantonese: ./configure --with-extdict-cmn && make There's also a distincion between latin text as pinyin/jyutping and latin text as English. So, these two should produce different results: espeak-ng -v cmn "english text 你好 more english text" -x espeak-ng -v cmn-Latn-pinyin "english text 你好 more english text" -x

Thanks for your help.
espeak version is 1.50
espeak-ng -v yue or cmn "你好" worked very will, but 【speak-ng --compile=yue or cmn】in dictsource:
Error processing file 'zhy_rules': No such file or directory.
I change yue to other language like tr tt vi , it can be installed, so i guess is the yue and cmn data packet broken because of change name?

espeak-ng -v cmn "你好" -x
can be read that's make me confuse, as i install language package failed the chinese also can be read.
but the output is n'i35_| x'Au214_|  is that IPA or something else? How can I get Pinyin instead of this?


./configure --with-extdict-cmn && make
Partial output results:
make  all-am
make[1]: Entering directory '/home/bage/Downloads/espeak-ng-master'
  CXXLD    src/espeak-ng
/usr/bin/ld: src/.libs/libespeak-ng.so: undefined reference to `sync_espeak_terminated_msg'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:1227: src/espeak-ng] Error 1
make[1]: Leaving directory '/home/bage/Downloads/espeak-ng-master'
make: *** [Makefile:830: all] Error 2

**Comment by mynah15** [comment]
> The way that `$textmode` works is that the corresponding text gets placed into the source buffer, so it is as if the text contains those characters. That means that with that change, those entries will now be interpreted as English instead of pinyin.
> 
> It would be useful to have a `textmode_dictrules` optin or similar, allowing this to use a different ruleset for that -- i.e. to always use the pinyin rules in this case. That could be quite complex to support.



> The way that `$textmode` works is that the corresponding text gets placed into the source buffer, so it is as if the text contains those characters. That means that with that change, those entries will now be interpreted as English instead of pinyin.
> 
> It would be useful to have a `textmode_dictrules` optin or similar, allowing this to use a different ruleset for that -- i.e. to always use the pinyin rules in this case. That could be quite complex to support.

I don't know how you complete it, but the pronunciation between pinyin and IPA is different, we can use pinyin to read chinese accurate, but IPA sounds like dialect or even hard to understand. If I can help, please let me know.Thanks

**Comment by jaacoppi** [comment]
For version 1.50 the languages are called zh and zhy. Try to compile those. The error message you're getting implies some other problem but I'm notsure since you're able to compile other languages.

Pinyin output is not supported. The normal output is Kirshenbaum ("ASCII IPA". You can use --ipa for IPA output.

**Comment by mynah15** [comment]
> lem but I'm notsure sin

Thanks for your reply, I will try it

**Comment by jaacoppi** [comment]
I think the best way to solve this is to change mandarin behave like cantonese with textmode. However, there's a problem with the  [Chao tones](https://www.liquisearch.com/tone_letter/chao_tone_letters_ipa).

Cantonese uses tone symbols 1-7, but Mandarin currently uses the Chao tones (55, 214 and so on) in phsrouce/ph_cmn and src/libespeak-ng/dictionary.c BUT normal tone numbers in dictsource/cmn_*. That results in a situation where the tones in dictionary files are not recognized and would be skipped  if textmode is handled like in Cantonese. Ways for fixing this are:

1. replace normal tones with Chao tones in dictionary list files. This would have to be redone every time the list is updated.
2. replace Chao tones with normal tones in the source code. This would probably lose some nuances.
3. Add the normal tones to phsource/ph_cmn. I'm not sure about the implications.


I'd go with the 3rd option and solve and see what happens.

**Comment by jaacoppi** [comment]
The attached patch seems to work. Reviews, please.
[cmn_fixes.patch.txt](https://github.com/espeak-ng/espeak-ng/files/7655653/cmn_fixes.patch.txt)

**Comment by jaacoppi** [comment]
My patch doesn't fully fix this. I've been trying to debug the mandarin voice but can't find a way to fix it except reverting to commit d91b7cf . That would break the Latin chars as English and would be back to pinyin.

The strange thing is that Cantonese works as expected, and I don't know why. 

Should we revert? This would require informing NVDA people that cmn and yue work differently again.

**Comment by hgneng** [comment]
> 
I have tried to apply the patch and it seems work. Could you show more details why you think the patch doesn't fully fix the issue?

**Comment by hgneng** [comment]
> My patch doesn't fully fix this. 
@jaacoppi  I have tried to apply the patch and it seems work. Could you show more details why you think the patch doesn't fully fix the issue?

**Comment by hgneng** [comment]
> My patch doesn't fully fix this. 

@jaacoppi  I have tried to apply the patch and it seems work. Could you show more details why you think the patch doesn't fully fix the issue?

**Comment by hgneng** [comment]
@jaacoppi Could you please merge your patch to the master stream first even though it may not be a full fix? Because the current version of Chinese is totally unrecognizable. Chinese NVDA users have been affected by this issue for a long time. And Linux users should also be affected soon.

I created Cantonese voice for eSpeak more than 10 years ago. And I also helped to create the Mandarin voice. Let me know if I can help with the testing or dictionary editing work.

**Comment by jaacoppi** [comment]
Sorry, I'm out for travelling and can't do anything before september. I don't remember what the problem with the patch is. Too bad I didn't write it down. Try different mixes of Chinese characters, pinyin and English to see where the patch works or fails.

**Comment by hgneng** [comment]
> > > My patch doesn't fully fix this.
> > 
> > 
> > @jaacoppi I have tried to apply the patch and it seems work. Could you show more details why you think the patch doesn't fully fix the issue?
> 
> I have the same issue, would like to ask how to use this patch?

Sorry for my late reply. (I haven't received github email for this issue. Don't know why.) To apply the patch:
1. download and unzip the code.
2. cd espeak-ng and copy the patch file [cmn_fixes.patch.txt](https://github.com/espeak-ng/espeak-ng/files/7655653/cmn_fixes.patch.txt) in espeak-ng
3. patch -p1 <cmn_fixes.patch.txt
4. ./autogen.sh && ./configure && make
5. src/espeak-ng -vcmn '你好，hello world'

**Comment by hgneng** [comment]
> Sorry, I'm out for travelling and can't do anything before september. I don't remember what the problem with the patch is. Too bad I didn't write it down. Try different mixes of Chinese characters, pinyin and English to see where the patch works or fails.

Never mind. I will wait until september.

I find one issue with this patch. It reads English not quite correct. Here is the result with the latest code and patch. "cmn" voice seems split English sentence into isolated word to read. "a" is read as "ei51".
```
$ ./espeak-ng -x "ma3 is a horse"
m'A: Tr'i:; Iz a# h'O@s
$ ./espeak-ng -x -vyue "ma3 is a horse"
(en)ma1Tr'i:1_| I1z_| a#1_| h'O@1s(yue)_|
$ ./espeak-ng -x -vcmn "ma3 is a horse"
(en)ma55Tr'i:55_| I55z(cmn)_| 'ei51_| (en)hO@11s(cmn)_|
$ ./espeak-ng -x -vcmn-Latn-pinyin 'ma3 is a horse'
m'A21_| (en)I44z(cmn)_| 'ei51_| Xo11@r11so-11_|
```

However, I do tests in espeak-1.48.15. "zh" voice reads English as "cmn-Latn-pinyin". "cmn" voice in espeak-ng-1.50 also reads English as "cmn-Latn-pinyin".
```
$ espeak --version
eSpeak text-to-speech: 1.48.15  16.Apr.15  Data at: /usr/lib/x86_64-linux-gnu/espeak-data
$ espeak -x -vzh '你好吗，ma3 is a horse'
Full dictionary is not installed for 'zh'
 n'i35_| x'Au21_| mA44_|
 m'A21_| (en)I44z(zh)_| 'ei51_| xo11@r11so-11_|
```

So, on the issue of not reading English correctly, the patch doesn't make things worse even it is not perfect. It just not make "cmn"(Mandarin) voice read English as good as "yue"(Cantonese). In the older version of eSpeak, "cmn" voice hasn't read English as good as "yue".

**Comment by jaacoppi** [comment]
I have now reviewed my patch again. It looks like it's definately better than the current situation. If there's no objections I'll merge it.

**Comment by hgneng** [comment]
Thank you very much!! I have tested the latest code and it works. Chinese espeak-ng and NVDA users should benefit from this patch a lot!

**Comment by cary-rowen** [comment]
Hi, @jaacoppi 
If this commit fixes the bug, you can close #1028 and #1163

Grateful
